### PR TITLE
ARROW-3746: [Gandiva] [Python] Print list of functions registered with gandiva

### PR DESCRIPTION
**Pull request by pcmoritz**
I'm also making the iterators of the Function registry static, can you check if that's ok @praveenbingo and @pravindra

### CODE REVIEW

**Review by pravindra — COMMENTED** [review]
we'd like to keep native_function.h internal since it has implementation details like internal fn name, flags, holders ..

For the similar requirement in java (i.e getting list of registered Fns), we implemented the APIs in expression_registry.h/cc - this exposes the list of functions and the types supported in gandiva. can you use those ?

**Comment by pcmoritz** [comment]
Thanks, I made that change!

File: cpp/src/gandiva/expression_registry.h [hunk 1]
```diff
@@ -61,5 +61,9 @@ class ExpressionRegistry {
   static void AddArrowTypesToVector(arrow::Type::type& type, DataTypeVector& vector);
   std::unique_ptr<FunctionRegistry> function_registry_;
 };
+
+void GetRegisteredFunctionSignatures(
```
Comment (pravindra) [inline cpp/src/gandiva/expression_registry.h]: curious - why not 

std::vector<std::shared_ptr<FunctionSignature>>* signatures) GetRegisteredFunctions() ?

**Review by pravindra — APPROVED** [review]
lgtm. thanks !

File: cpp/src/gandiva/expression_registry.h (hunk 1 above)
Comment (pcmoritz) [reply cpp/src/gandiva/expression_registry.h]: In general I like the pattern where the client allocates the vector and the function pushes onto it, it allows adding elements to an existing vector. In this case I agree it's not necessary, I changed it to your suggestion.

**Review by pcmoritz — COMMENTED** [review]
+1 can be merged when the tests pass

**Pull request closed without merging**